wallet/src/Pos/Util/Mnemonic.hs

```diff
@@ -0,0 +1,664 @@
+        "<mnemonic>"
+
+-- | To use everytime we need to show an example of a Mnemonic. This particular
+-- mnemonic is rejected to prevent users from using it on a real wallet.
+instance Default Mnemonic where
+    def = Mnemonic
+        { getEntropy = Entropy "\211\177\US\"\245\163\233\152\169]\149\DC3\239)\234\172"
+        , getWords =
+            [ (1693,"squirrel")
+            , (1095,"material")
+            , (1605,"silly")
+            , (1882,"twice")
+            , (500,"direct")
+            , (1634,"slush")
+            , (1323,"pistol")
+            , (1429,"razor")
+            , (159,"become")
+            , (970,"junk")
+            , (981,"kingdom")
+            , (710,"flee")
+            ]
+        }
+
+-- Same remark from 'Arbitrary Entropy' applies here.
+instance Arbitrary Mnemonic where
+    arbitrary =
+        entropyToMnemonic <$> arbitrary
+
+instance FromJSON Mnemonic where
+    parseJSON =
+        parseJSON >=> (eitherToParser . mkMnemonic)
+      where
+        eitherToParser :: Either Text Mnemonic -> Parser Mnemonic
+        eitherToParser =
+            either (fail . show) pure
+
+instance ToJSON Mnemonic where
+    toJSON =
+        toJSON . map snd . getWords
+
+instance ToSchema Mnemonic where
+    declareNamedSchema _ = do
+        NamedSchema _ schema <- declareNamedSchema (Proxy @[Text])
+        return $ NamedSchema (Just "Mnemonic") schema
+
+
+-- | Smart-constructor for the Entropy
+mkEntropy :: ByteString -> Either Text Entropy
+mkEntropy bytes = do
+    when (BS.length bytes <= 0) $
+        Left "Entropy must be a non-zero sequence of bytes"
+    when (BS.length bytes > 64) $
+        Left "Maximum entropy is 64 bytes (512 bits)"
+    when (BS.length bytes `rem` 4 /= 0) $
+        Left "Entropy must be a multiple of 4 bytes"
+    Right (Entropy bytes)
+
+
+-- | Smart-constructor for the Mnemonic
+mkMnemonic :: [Text] -> Either Text Mnemonic
+mkMnemonic wordsm = do
+    when (length wordsm <= 0) $
+        Left "Mnemonic must be a non-empty list of words"
+    when (length wordsm > 48) $
+        Left "Mnemonic must have a maximum number of 48 words"
+    when (length wordsm `rem` 3 /= 0) $
+        Left "Mnemonic must be a multiple of 3 words"
+    when (isJust $ Text.find (not . isAscii) (unwords wordsm)) $
+        Left "Mnemonic must use only ASCII characters"
+    when (wordsm == map snd (getWords def)) $
+        Left "Forbidden Mnemonic: an example Mnemonic has been submitted. Please generate a fresh and private Mnemonic from a trusted source"
+    indices <- findIndices enWordList wordsm
+    let bits = concatMap word11ToBits indices
+    let (entropyBits, checksum) = splitAt ((length bits `quot` 33) * 32) bits
+    entropy <- mkEntropy (bitsToBytes entropyBits)
+    when (checksum /= calcChecksum entropy) $
+        Left "mnemonic checksum failed: cannot convert mnemonic to entropy"
+    pure $ Mnemonic
+        { getEntropy = entropy
+        , getWords   = zip indices wordsm
```

Comment:
Those are hidden from outside of the module, I put them in the Mnemonic object directly to avoid having to re-compute them when doing conversions. Since Mnemonic can only be created using the public functions of the module, it's safe (no one has access to `getWords` or `indices` in the end).